src/wrapper/java_vm/vm.rs

```diff
@@ -132,9 +190,129 @@ impl<'a> Deref for AttachGuard<'a> {
 }
 
 impl<'a> Drop for AttachGuard<'a> {
+    fn drop(&mut self) {
+        if self.should_detach {
+            InternalAttachGuard::clear_tls();
+        }
+    }
+}
+
+#[derive(PartialEq)]
+enum ThreadType {
+    Normal,
+    Daemon,
+}
+
+#[derive(Debug)]
+struct InternalAttachGuard {
+    java_vm: *mut sys::JavaVM,
+}
+
+impl InternalAttachGuard {
+    fn create_and_attach(
+        java_vm: *mut sys::JavaVM,
+        ty: ThreadType,
+    ) -> Result<*mut sys::JNIEnv> {
+        let guard = InternalAttachGuard::new(java_vm);
+        let env_ptr = unsafe {
+            if ty == ThreadType::Daemon {
+                guard.attach_current_thread_as_daemon()?
+            } else {
+                guard.attach_current_thread()?
+            }
+        };
+
+        Self::fill_tls(guard);
+
+        Ok(env_ptr)
+    }
+
+    fn new(java_vm: *mut sys::JavaVM) -> Self {
+        Self {
+            java_vm,
+        }
+    }
+
+    /// Stores guard in thread local storage.
+    fn fill_tls(guard: InternalAttachGuard) {
+        THREAD_ATTACH_GUARD.with(move |f| {
+            *f.borrow_mut() = Some(guard);
+        });
+    }
+
+    /// Clears thread local storage, dropping the InternalAttachGuard and causing detach of
+    /// the current thread.
+    fn clear_tls() {
+        THREAD_ATTACH_GUARD.with(move |f| {
+            *f.borrow_mut() = None;
+        });
+    }
+
+    unsafe fn attach_current_thread(&self) -> Result<*mut sys::JNIEnv> {
+        let mut env_ptr = ptr::null_mut();
+        let res = java_vm_unchecked!(
+            self.java_vm,
+            AttachCurrentThread,
+            &mut env_ptr,
+            ptr::null_mut()
+        );
+        jni_error_code_to_result(res)?;
+
+        ATTACHED_THREADS.fetch_add(1, Ordering::SeqCst);
```

Comment:
Maybe it makes sense to use value returned by `fetch_add`?

Reply:
It returns previous value, which is not much useful for us. We can increment it, but I think it isn't worth the effort.